### ISSUE TITLE
Using a Foreign Event loop with a client websocket (outbound) connection

### ISSUE THREAD

**Issue by emusgrave** [warmcat/libwebsockets#2108]
Does the [Minimal HTTP Client Multi](https://github.com/warmcat/libwebsockets/blob/main/minimal-examples/http-client/minimal-http-client-multi/minimal-http-client-multi.c) example really show the correct way to use a foreign event loop as of v4.x? It is setting the `LWS_SERVER_OPTION_LIBUV` flag, but then it still uses `lws_service` instead of setting the `creation_info.foreign_loops` parameter.

I'm new to the library, but I was able to setup a simple outgoing websocket client connection using the built in event loop (with `lws_service`). However, when I try integrating into a foreign libuv event loop, I'm unable to make a connection to the server. I am getting either an `LWS_CALLBACK_CLIENT_CONNECTION_ERROR` with the message "Unable to connect", or a `I: lws_client_conn_wait_timeout: connect wait timeout has fired` depending on my order of operations. (I've tried many different combinations of creating the context before starting the libuv loop with `uv_run`, to creating the context in a timer after the event loop is already running, etc)

I can setup a simple example to demonstrate the issue, but first I wanted to pose the question if I'm just overlooking that the `lws_service` call is still required or if there is something else that is not documented for the client side to use a foreign event loop.

**Comment by lws-team** [comment]
> Does the Minimal HTTP Client Multi example really show the correct way to use a foreign event loop as of v4.x?

No... does it say anywhere that it is trying to "show the correct way to use a foreign event loop"?  It just calls itself minimal-http-client-multi, meaning multiple client connections.

The event loop doesn't know or care if you are listening or making client connections or some mix, they are all wsi with fds that have watchers on the event loop.  So start from the example that is dedicated to showing how to handle foreign loops

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-server/minimal-http-server-eventlib-foreign

and add on the client creation pieces.

**Comment by emusgrave** [comment]
Sorry, I didn't realize that setting `LWS_SERVER_OPTION_LIBUV` could be done separate from a foreign event loop, so my searching for examples of using foreign loops with a client connection was coming up with false positives.

I've now tried taking the minimal-http-server-eventlib-foreign and converting it to make client connection, but still can't get it to work. I also tried taking the minimal-ws-client-spam and converting it to use a foreign loop, and get to the same point:

Once I try to make the connection I get:
```
lws_client_conn_wait_timeout: connect wait timeout has fired
lws_client_connect_3_connect: abandoning connect due to timeout
```
This happens right after the callback `LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED` is fired.

I should also mention that I'm trying this on Windows. My next step will be to spin up a linux box and give it a try there.

**Comment by emusgrave** [comment]
@lws-team I can confirm that a client connection w/foreign libuv loop works fine on Linux, but does not work on Windows (same code, compiling cross-platform).

I realize Windows is probably not a priority for you. If I get some free time I'll see if I can debug into libwebsockets on Windows and track down the issue, but for now I need to move forward so I am going to stick with Linux.

**Comment by lws-team** [comment]
What breaks in the windows + libuv case?

**Comment by emusgrave** [comment]
After attempting to kick off the connection with `lws_client_connect_via_info`, it receives the `LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED`, but then gets 
```
lws_client_conn_wait_timeout: connect wait timeout has fired
lws_client_connect_3_connect: abandoning connect due to timeout
```
Whereas the working case on linux will get a callback of `LWS_CALLBACK_EVENT_WAIT_CANCELLED` at that same point, and then continue connecting. I'm connecting to echo.websocket.org and it's using the identical source code in both cases, so it doesn't seem to be a server-side problem.

Here are some log snippets that compare the two cases:

### Working (Linux)
```
[2020/11/10 15:01:46:9312] D: lws_client_connect_3_connect: 0x555555a4cc60: WAITING_CONNECT
[2020/11/10 15:01:46:9312] D: lwsi_set_state(0x555555a4cc60, 0x10000202)
[2020/11/10 15:01:46:9312] D: _realloc: size 160: uvh
[2020/11/10 15:01:46:9312] D: __insert_wsi_socket_into_fds: 0x555555a4cc60: tsi=0, sock=11, pos-in-fds=1
[2020/11/10 15:01:46:9312] D: elops_io_uv: 0x555555a4cc60: 5
[2020/11/10 15:01:46:9312] D: _lws_change_pollfd: wsi 0x555555a4cc60: fd 11 events 1 -> 1
[2020/11/10 15:01:46:9312] D: elops_io_uv: 0x555555a4cc60: 5
[2020/11/10 15:01:46:9312] D: __lws_set_timeout: 0x555555a4cc60: 20 secs, reason 2
[2020/11/10 15:01:46:9314] D: lws_client_connect_3_connect: connect says errno: 115
[2020/11/10 15:01:46:9314] D: _lws_change_pollfd: wsi 0x555555a4cc60: fd 11 events 1 -> 5
[2020/11/10 15:01:46:9314] D: elops_io_uv: 0x555555a4cc60: 6
[2020/11/10 15:01:46:9314] D: lws_client_connect_via_info: wsi 0x555555a4cc60: adoption cb 19 to h1 lws-spam-test
[2020/11/10 15:01:46:9315] U: LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED
[2020/11/10 15:01:46:9315] D: lws_cancel_service
[2020/11/10 15:01:46:9315] D: _lws_state_transition: system: changed 7 'CPD_POST_TIME' -> 8 'POLICY_VALID'
[2020/11/10 15:01:46:9315] D: _lws_state_transition: system: changed 8 'POLICY_VALID' -> 9 'REGISTERED'
[2020/11/10 15:01:46:9315] D: _lws_state_transition: system: changed 9 'REGISTERED' -> 10 'AUTH1'
[2020/11/10 15:01:46:9315] D: _lws_state_transition: system: changed 10 'AUTH1' -> 11 'AUTH2'
[2020/11/10 15:01:46:9315] D: _lws_state_transition: system: changed 11 'AUTH2' -> 12 'OPERATIONAL'
[2020/11/10 15:01:46:9315] I: lws_state_transition_steps: CONTEXT_CREATED -> OPERATIONAL
[2020/11/10 15:01:46:9315] U: LWS_CALLBACK_EVENT_WAIT_CANCELLED
```

### Not Working (Windows)
```
[2020/11/10 14:56:26:9438] D: lws_client_connect_3_connect: 000001E402584C90: WAITING_CONNECT
[2020/11/10 14:56:26:9468] D: lwsi_set_state(000001E402584C90, 0x10000202)
[2020/11/10 14:56:26:9478] D: _realloc: size 416: uvh
[2020/11/10 14:56:26:9498] D: __insert_wsi_socket_into_fds: 000001E402584C90: tsi=0, sock=948, pos-in-fds=0
[2020/11/10 14:56:26:9518] D: _lws_change_pollfd: wsi 000001E402584C90: fd 948 events 9 -> 41
[2020/11/10 14:56:26:9538] D: elops_io_uv: 000001E402584C90: 5
[2020/11/10 14:56:26:9558] D: __lws_set_timeout: 000001E402584C90: 20 secs, reason 2
[2020/11/10 14:56:26:9578] D: lws_client_connect_3_connect: connect says errno: 10035
[2020/11/10 14:56:26:9598] D: lws_client_connect_via_info: wsi 000001E402584C90: adoption cb 19 to h1 lws-spam-test
[2020/11/10 14:56:26:9648] U: LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED
[2020/11/10 14:56:26:9668] D: lws_cancel_service
[2020/11/10 14:56:26:9698] D: _lws_state_transition: system: changed 7 'CPD_POST_TIME' -> 8 'POLICY_VALID'
[2020/11/10 14:56:26:9718] D: _lws_state_transition: system: changed 8 'POLICY_VALID' -> 9 'REGISTERED'
[2020/11/10 14:56:26:9738] D: _lws_state_transition: system: changed 9 'REGISTERED' -> 10 'AUTH1'
[2020/11/10 14:56:26:9767] D: _lws_state_transition: system: changed 10 'AUTH1' -> 11 'AUTH2'
[2020/11/10 14:56:26:9787] D: _lws_state_transition: system: changed 11 'AUTH2' -> 12 'OPERATIONAL'
[2020/11/10 14:56:26:9807] I: lws_state_transition_steps: CONTEXT_CREATED -> OPERATIONAL
[2020/11/10 14:56:32:1164] I: lws_client_conn_wait_timeout: connect wait timeout has fired
[2020/11/10 14:56:32:1184] I: lws_client_connect_3_connect: abandoning connect due to timeout
[2020/11/10 14:56:32:1204] D: elops_io_uv: 000001E402584C90: -2147483637
[2020/11/10 14:56:32:1254] U: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2020/11/10 14:56:32:1274] E: CLIENT_CONNECTION_ERROR: Unable to connect (try 1, est 0, closed 0, err 1)
```

**Comment by lws-team** [comment]
Basically for windows there's no indication coming from libuv `uv_poll_t` that the asynchonous connect() succeeded or failed.

With POSIX sockets, pending async connect()s will signal POLLOUT event on the fd when there is a definitive result for the `connect()` available, so we wait for that whatever the event lib.  As you found that works as expected... with POSIX sockets implementations like Linux.

With microsoft's WSA, as I unfortunately know from making the default lws loop work with it, there's a non-POSIX connect()-specific event mask `FD_CONNECT_BIT` instead.  On windows, it also has been made to work as expected for the default lws platform loop, based on WSA for windows platform.

With windows + libuv, most everything else is made to fit into POSIX semantics by either microsoft or libuv, but rather than fake up the POLLOUT / connect behaviour in `uv_poll_t` they wrap the windows proprietary nonsense in different libuv abstractions `uv_tcp_t` / `uv_udp_t` and `uv_connect_t` that we don't use or otherwise need.  I have opened an issue on libuv about the lack of indication on windows platform for client connect completion using their uv_poll_t apis https://github.com/libuv/libuv/issues/3045

Unfortunately if you wrap the fd in one of these uv_tcp_t there's no way to detach the fd again so you can continue with uv_poll_t after the connect completed, which would have been fine, closing the uv_tcp_t closes the underlying fd, and while the uv_tcp_t exists you can't re-register the same fd using uv_poll_t.

Unless lws libuv support is rewritten to use a mixture of the existing stuff (for file descriptor events), `uv_tcp_t` and `uv_udp_t`, to no real gain except fixing this, or libuv consider sythesizing the native FD_CONNECT event as a POLLOUT to match POSIX semantics in a crossplatform way, for libuv + windows client it looks like it needs a periodic polling type check to see if the connect went through.

**Comment by vtjnash** [comment]
> there's no way to detach the fd again

There's trivially `dup(uv_fileno())` (don't forget to set CLOEXEC too!). Though I recently noticed there is a problem with it on Windows, but I'm currently in the process of fixing that (https://github.com/libuv/libuv/pull/3036)

**Comment by lws-team** [comment]
@vtjnash Thanks for the tip

> There's trivially

"Trivial" for me would look like libuv folding POSIX select event behaviour into libuv windows platform support for `uv_poll_t`.

It doesn't really look like creating a second, parallel, different temporary event library socketfd abstraction object to own the fd to catch that event only, just on the one platform, then get out of the nondetachable fd  fingertrap by whipping out a dup() I prepared earlier for that purpose, and kill the temp object.  For every client connection  It looks like an ugly platform-specific workaround (not getting into the uv_poll_t vs other abstraction schism in libuv).

libuv's offer is to simplify life by concealing the platform differences...
